hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/utils/AutoCompletion.java

```diff
@@ -0,0 +1,164 @@
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.ozone.utils;
+
+import org.apache.hadoop.hdds.scm.server.StorageContainerManagerStarter;
+import org.apache.hadoop.ozone.HddsDatanodeService;
+import org.apache.hadoop.ozone.admin.OzoneAdmin;
+import org.apache.hadoop.ozone.audit.parser.AuditParser;
+import org.apache.hadoop.ozone.conf.OzoneGetConf;
+import org.apache.hadoop.ozone.csi.CsiServer;
+import org.apache.hadoop.ozone.debug.OzoneDebug;
+import org.apache.hadoop.ozone.freon.Freon;
+import org.apache.hadoop.ozone.genconf.GenerateOzoneRequiredConfigurations;
+import org.apache.hadoop.ozone.insight.Insight;
+import org.apache.hadoop.ozone.om.OzoneManagerStarter;
+import org.apache.hadoop.ozone.recon.ReconServer;
+import org.apache.hadoop.ozone.repair.OzoneRepair;
+import org.apache.hadoop.ozone.s3.Gateway;
+import org.apache.hadoop.ozone.shell.OzoneRatis;
+import org.apache.hadoop.ozone.shell.OzoneShell;
+import org.apache.hadoop.ozone.shell.checknative.CheckNative;
+import org.apache.hadoop.ozone.shell.s3.S3Shell;
+import org.apache.hadoop.ozone.shell.tenant.TenantShell;
+import picocli.AutoComplete;
+import picocli.CommandLine;
+import picocli.CommandLine.Option;
+
+/**
+ * Tool to generate bash/zsh auto-complete scripts for Ozone CLI.
+ */
+public final class AutoCompletion {
+
+  private AutoCompletion() { }
+
+  public static void main(String[] args) {
+
+    CommandLine hierarchy = new CommandLine(new Ozone())
+        .addSubcommand("admin", new OzoneAdmin().getCmd())
+        .addSubcommand("auditparser", new AuditParser().getCmd())
+        .addSubcommand("checknative", new CheckNative().getCmd())
+        .addSubcommand("csi", new CsiServer().getCmd())
+        .addSubcommand("datanode", new HddsDatanodeService().getCmd())
+        .addSubcommand("debug", new OzoneDebug().getCmd())
+        .addSubcommand("freon", new Freon().getCmd())
+        .addSubcommand("genconf", new GenerateOzoneRequiredConfigurations().getCmd())
+        .addSubcommand("getconf", new OzoneGetConf().getCmd())
+        .addSubcommand("httpfs", new HttpFSServerWebServer())
+        .addSubcommand("insight", new Insight().getCmd())
+        .addSubcommand("om", new OzoneManagerStarter(null).getCmd())
+        .addSubcommand("ratis", new OzoneRatis().getCmd())
+        .addSubcommand("recon", new ReconServer().getCmd())
+        .addSubcommand("repair", new OzoneRepair().getCmd())
+        .addSubcommand("s3", new S3Shell().getCmd())
+        .addSubcommand("s3g", new Gateway().getCmd())
+        .addSubcommand("scm", new StorageContainerManagerStarter(null).getCmd())
+        .addSubcommand("sh", new OzoneShell().getCmd(), "shell")
+        .addSubcommand("tenant", new TenantShell().getCmd())
+        .addSubcommand("version", new OzoneVersionInfo());
+
+    /*
+     * The following commands are marked as hidden,
+     * so they won't have auto complete.
+     *   - OzoneAdmin
```

Comment:
I think `OzoneAdmin` can be made visible.  It's no more dangerous than `OzoneRepair`.

Reply:
+1, will do.